File: 5/25.k-个一组翻转链表.py
#
# @lc app=leetcode.cn id=25 lang=python3
#
# [25] K 个一组翻转链表
#
# https://leetcode-cn.com/problems/reverse-nodes-in-k-group/description/
#
# algorithms
# Hard (66.70%)
# Likes:    1584
# Dislikes: 0
# Total Accepted:    315.6K
# Total Submissions: 472.5K
# Testcase Example:  '[1,2,3,4,5]\n2'
#
# 给你一个链表，每 k 个节点一组进行翻转，请你返回翻转后的链表。
# 
# k 是一个正整数，它的值小于或等于链表的长度。
# 
# 如果节点总数不是 k 的整数倍，那么请将最后剩余的节点保持原有顺序。
# 
# 进阶：
# 
# 
# 你可以设计一个只使用常数额外空间的算法来解决此问题吗？
# 你不能只是单纯的改变节点内部的值，而是需要实际进行节点交换。
# 
# 
# 
# 
# 示例 1：
# 
# 
# 输入：head = [1,2,3,4,5], k = 2
# 输出：[2,1,4,3,5]
# 
# 
# 示例 2：
# 
# 
# 输入：head = [1,2,3,4,5], k = 3
# 输出：[3,2,1,4,5]
# 
# 
# 示例 3：
# 
# 
# 输入：head = [1,2,3,4,5], k = 1
# 输出：[1,2,3,4,5]
# 
# 
# 示例 4：
# 
# 
# 输入：head = [1], k = 1
# 输出：[1]
# 
# 
# 
# 
# 
# 提示：
# 
# 
# 列表中节点的数量在范围 sz 内
# 1 
# 0 
# 1 
# 
# 
#


from commons.List import ListNode
from typing import Optional


# @lc code=start
class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if not head or k <= 1:
            return head

        dummy = ListNode(-1, head)

        begin = curr = head
        prev_tail = dummy

        while curr:
            prev = None
            begin = curr

            for _ in range(k):
                if not curr:
                    curr = prev
                    prev = None

                    while curr:
                        next = curr.next
                        curr.next = prev
                        prev = curr
                        curr = next
                    break

                next = curr.next
                curr.next = prev
                prev = curr
                curr = next

            prev_tail.next = prev
            prev_tail = begin

        return dummy.next
# @lc code=end

File: 25.k-个一组翻转链表.ts
/*
 * @lc app=leetcode.cn id=25 lang=typescript
 *
 * [25] K 个一组翻转链表
 *
 * https://leetcode.cn/problems/reverse-nodes-in-k-group/description/
 *
 * algorithms
 * Hard (67.64%)
 * Likes:    1759
 * Dislikes: 0
 * Total Accepted:    383.3K
 * Total Submissions: 566.6K
 * Testcase Example:  '[1,2,3,4,5]\n2'
 *
 * 给你链表的头节点 head ，每 k 个节点一组进行翻转，请你返回修改后的链表。
 *
 * k 是一个正整数，它的值小于或等于链表的长度。如果节点总数不是 k 的整数倍，那么请将最后剩余的节点保持原有顺序。
 *
 * 你不能只是单纯的改变节点内部的值，而是需要实际进行节点交换。
 *
 *
 *
 * 示例 1：
 *
 *
 * 输入：head = [1,2,3,4,5], k = 2
 * 输出：[2,1,4,3,5]
 *
 *
 * 示例 2：
 *
 *
 *
 *
 * 输入：head = [1,2,3,4,5], k = 3
 * 输出：[3,2,1,4,5]
 *
 *
 *
 * 提示：
 *
 *
 * 链表中的节点数目为 n
 * 1 <= k <= n <= 5000
 * 0 <= Node.val <= 1000
 *
 *
 *
 *
 * 进阶：你可以设计一个只用 O(1) 额外内存空间的算法解决此问题吗？
 *
 *
 *
 *
 */

import { ListNode } from './commons/list'

export
// @lc code=start
function reverseKGroup(head: ListNode | null, k: number): ListNode | null {
  const dummy = new ListNode(-1, head)

  let prevEnd = dummy
  let curr = dummy.next

  while (curr) {
    let prev: ListNode | null = null
    let insufficientNode = false
    const groupEntrance = curr

    for (let i = 0; i < k; i++) {
      if (!curr) {
        insufficientNode = true
        break
      }
      [prev, curr] = reverse(prev, curr)
    }

    if (insufficientNode) {
      curr = prev
      prev = null
      while (curr) {
        [prev, curr] = reverse(prev, curr)
      }
    }

    prevEnd.next = prev
    if (!insufficientNode) {
      groupEntrance.next = curr
    }

    prevEnd = groupEntrance
  }

  return dummy.next
}

function reverse(prev: ListNode | null, curr: ListNode) {
  const next = curr.next
  curr.next = prev
  return [curr, next]
}
// @lc code=end
